### PR TITLE
12.0 fix pref pos pim

### PR DESCRIPTION
**Pull request by pimodoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pimodoo** [comment]
Looks good to you?

**Comment by pimodoo** [comment]
@kebeclibre

File: addons/point_of_sale/models/pos_order.py [hunk 1]
```diff
@@ -118,19 +118,25 @@ def _process_order(self, pos_order):
...
-
         if pos_session.sequence_number <= pos_order['sequence_number']:
             pos_session.write({'sequence_number': pos_order['sequence_number'] + 1})
             pos_session.refresh()
+        return order
 
-        if not float_is_zero(pos_order['amount_return'], prec_acc):
-            cash_journal_id = pos_session.cash_journal_id.id
+    @api.multi
+    def _get_payments(self,data):
```
Comment (kebeclibre) [inline addons/point_of_sale/models/pos_order.py]: docstring data is actually a dict representing an order

File: addons/point_of_sale/models/pos_order.py [hunk 2]
```diff
@@ -721,16 +728,22 @@ def create_from_ui(self, orders):
...
                 self._match_payment_to_invoice(order)
             pos_order = self._process_order(order)
-            order_ids.append(pos_order.id)
+            order_ids |= pos_order
+            prepared_payments.append(pos_order._get_payments(order))
+
+        self.env['pos.order']._create_payments(prepared_payments)
 
+        for pos_order in order_ids:
+            pos_order.amount_paid = sum(payment.amount for payment in pos_order.statement_ids)
```
Comment (kebeclibre) [inline addons/point_of_sale/models/pos_order.py]: remove this line

File: addons/point_of_sale/models/pos_order.py [hunk 3]
```diff
@@ -943,13 +955,19 @@ def _prepare_bank_statement_line_payment_values(self, data):
 
         return args
 
+    @api.model
+    def _create_payments(self, payments):
```
Comment (kebeclibre) [inline addons/point_of_sale/models/pos_order.py]: docstring payment is list of dict

File: addons/point_of_sale/models/pos_order.py [hunk 4]
```diff
@@ -839,7 +852,6 @@ def create_picking(self):
     def _force_picking_done(self, picking):
         """Force picking in order to be set as done."""
         self.ensure_one()
-        picking.action_assign()
```
Comment (kebeclibre) [inline addons/point_of_sale/models/pos_order.py]: a priori ok for this
I'd like to see a test though

File: addons/account/models/account_bank_statement.py [hunk 5]
```diff
@@ -340,7 +340,7 @@ def _check_currency_id(self):
 
     @api.model
```
Comment (kebeclibre) [inline addons/account/models/account_bank_statement.py]: create_multi

**Comment by kebeclibre** [comment]
The art of the deal:
- coverage tests and verify output, obviously
- functional test on the picking

File: addons/point_of_sale/models/pos_order.py [hunk 6]
```diff
@@ -730,32 +730,37 @@ def create_from_ui(self, orders):
...
+        orders_to_invoice = self.env['pos.order']
         for tmp_order in orders_to_save:
             to_invoice = tmp_order['to_invoice']
             order = tmp_order['data']
             if to_invoice:
                 self._match_payment_to_invoice(order)
+
             pos_order = self._process_order(order)
             order_ids |= pos_order
+            if to_invoice:
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_order.py]: you already have if to_invoice up

File: addons/point_of_sale/models/pos_order.py [hunk 7]
```diff
@@ -835,27 +844,34 @@ def create_picking(self):
...
-    def _force_picking_done(self, picking):
+    def _force_picking_done(self, picking_dict):
         """Force picking in order to be set as done."""
-        self.ensure_one()
-        picking.action_assign()
-        wrong_lots = self.set_pack_operation_lot(picking)
-        if not wrong_lots:
-            picking.action_done()
+        pickings_to_assign = self.env['stock.picking']
+        for order_id, picking in picking_dict.items():
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_order.py]: do we really need to assign before done, or can we imediattely done?
If we can if not wrong_lots -> action_done, else -> action assign

If we need to assign, remove the loop and do picking_to_assign = self.env['stock.picking'].concat(*picking_dict.values())

**Comment by mart-e** [comment]
Is this one still needed or leftover from previous dev?

**Comment by pimodoo** [comment]
@mart-e it was for @tdi-odoo but never used.

**Pull request closed without merging**